### PR TITLE
[HW] Support type aliases where ArrayType is required.

### PR DESCRIPTION
**Pull request by mikeurbach**
The initial commit's comment is no longer valid... we have discussed and prototyped several approaches here.

### CODE REVIEW

**Comment by mikeurbach** [comment]
Here is the support for using type aliases in the array* operations. If this makes sense, I will add the same support for the struct* operations. After that, there should be no more ODS changes to support type aliases and declarations.

File: include/circt/Dialect/HW/HWTypes.td
Comment (teqdruid) [inline include/circt/Dialect/HW/HWTypes.td:174]: Parsing this is making me dizzy... Can you add newlines and indents to help with the nesting?

File: lib/Dialect/HW/HWTypes.cpp
Comment (teqdruid) [inline lib/Dialect/HW/HWTypes.cpp:152]: I think `cast` will assert on fail, which you don't want.

**Review by teqdruid — APPROVED** [review]
LGTM. Thanks!

File: lib/Dialect/HW/HWTypes.cpp
Comment (mikeurbach) [reply lib/Dialect/HW/HWTypes.cpp:152]: We know statitically that if it is a type alias, then its canonical type is an array type. I was thinking about this, and I think I will just make TypeAliasOr a proper TypeDef, so this casting can be a generic `getCanonicalType` on `TypeAliasOr`, rather than having this (and other) free functions to do the cast.

Comment (mikeurbach) [reply lib/Dialect/HW/HWTypes.cpp:152]: Actually, I don't think subclassing TypeDef is what we want... that would add another layer of wrapping the type during parsing and printing, when really we just want to represent that it can parse/print as either an alias or the inner type, and have a method to canonicalize to the inner type when neccessary. So what we really want is a DialectType, with an extra helper method for casting... that's basically what this is. I don't know if there is a way to do this and attach the helper to the DialectType rather than just using a free function like this.

Comment (mikeurbach) [reply lib/Dialect/HW/HWTypes.cpp]: For now I've made it a DialectType and moved `cast`->`dyn_cast` so this returns null rather than asserting in case the assumption is not met. I also copied the comment about this behavior to the implementation file as well to make it clear.

I'm still curious if it would be possible to attach the `castArrayType` functionality onto the DialectType itself, but I think the only way to do that would be to add a new Type subclass and use it for the DialectType. I can explore that, but it may be more trouble and indirection than its worth.

Comment (mikeurbach) [reply lib/Dialect/HW/HWTypes.cpp]: > I can explore that

It's not actually that bad... I will update this PR

**Comment by mikeurbach** [comment]
Struct / Union support staged here as a follow on to this approach: https://github.com/llvm/circt/pull/1150

File: include/circt/Dialect/HW/HWTypes.h
Comment (teqdruid) [inline include/circt/Dialect/HW/HWTypes.h:26]: This approach is neat. I'd like @lattner to weigh in on what he thinks... I'm concerned that there are caveats/subtleties  involved that could cause this to fail in some cases.

Comment (teqdruid) [reply include/circt/Dialect/HW/HWTypes.h:26]: Also, if this is appropriate there are advantages to having this declared in ODS. Non-parameterized Typedefs should be supported...

Comment (mikeurbach) [reply include/circt/Dialect/HW/HWTypes.h:26]: @teqdruid this is sort of declared in ODS. The ODS TypeAliasOr type is backed by this C++ TypeAliasOr. Did you mean to declare it as a TypeDef instead of a DialectType? I don't think that's desirable for the reason [mentioned](https://github.com/llvm/circt/pull/1143#discussion_r638900259) above.

I'm not sure what you mean about supporting non-parameterized Typedefs... We already support TypeAliasType in most of the HW dialect types (HWValueType, etc.), we just need to support the cases where an op like array_get previously expected just an ArrayType, and we now want to also support aliases there.

Comment (teqdruid) [reply include/circt/Dialect/HW/HWTypes.h:26]: Typedefs in ODS don't necessarily involve printing/parsing/wrapping. I _think_ it could be used to generate what you've got, but with the additional advantage of being able to use it instead of the type constraints you're writing by hand. I'm imagining a tblgen `class` (which extends typedef) then `def`s for array, struct, and union.

Except for the `classof` override, which is the whole point... Though you could put that in the `extraDeclarations` section.

I'm not particularly opposed to this C++ implementation, though that `classof` override may have implications of which we are not aware.

File: include/circt/Dialect/HW/HWTypes.h
Comment (darthscsi) [inline include/circt/Dialect/HW/HWTypes.h:47]: so if I have:
```
typedef i64 a;
typedef a[8] b;
b foo;
```
what is the type of foo[1]?
is it a b or an i64?  It looks like here it would be an i64.

File: include/circt/Dialect/HW/HWTypes.h
Comment (mikeurbach) [reply include/circt/Dialect/HW/HWTypes.h:26]: > (which extends typedef)

This is what I initially tried [yesterday](https://github.com/llvm/circt/pull/1143#discussion_r638900259) before deciding a `DialectType` with a custom backing C++ class was more appropriate. I can take another look, but I'm still not sure what would be gained.

As it stands, the ODS `TypeAliasOrConstraint` and `TypeAliasOr` class work together with the C++ `TypeAliasOr` class to transparently let everything (e.g. the parsers, where `parseType` still just works) work with either a `TypeAliasType` or some other type. In other words, there is no way to represent `TypeAliasOr` in the IR, which I think is exactly what we want, and why I chose a `DialectType` rather than a `TypeDef`. My understanding is you would use a TypeDef if you want to have a new type that could be represented in the IR, but that's not really what we want, we just need a "union type" from the perspective of the verifiers, printers, and parsers. Does that make sense? Is there some benefit to using a `TypeDef` here I'm missing?

> though that classof override may have implications of which we are not aware.

`classof` is the hook to customize `isa`,`cast`,`dyn_cast` functionality: https://github.com/llvm/llvm-project/blob/969eefd98e0f8e485148be61190cc2ef62fb1eca/mlir/include/mlir/IR/Types.h#L228

I just checked the MLIR codebase, and indeed this hook is only called by the various implementations of `isa` (and a couple places in the standard dialect / generated code that should probably just call `isa`). So I override it to create the "union type"-like behavior described above, which seems like a valid use case (and empirically works with all existing verifiers, parsers, and printers). Can you be more specific about your concerns? I'm not sure what other implication there could be.

**Review by darthscsi — APPROVED** [review]
I like the constraints.  My only concern is deciding how nesting of typedefs work with respect to indexing structs and arrays.

File: include/circt/Dialect/HW/HWTypes.h
Comment (teqdruid) [reply include/circt/Dialect/HW/HWTypes.h:26]: I get what you're doing and why. This just doesn't seem like a standard use case to me, so I'm concerned about unknown unknowns, so I'd feel more comfortable if Chris said it was fine. You might bring it up at today's ODM just to get Chris' blessing on this one small thing.

That said, since it works I have to objections to merging it.

Comment (mikeurbach) [reply include/circt/Dialect/HW/HWTypes.h:26]: Sure thing, understood about "unknown unknowns".

File: include/circt/Dialect/HW/HWTypes.h
Comment (darthscsi) [reply include/circt/Dialect/HW/HWTypes.h:47]: On more thought, as long as the types compare equal, either should be valid and could just be a function of what the builder was called with.  So if you want the sugar type, you construct the op with the result being the sugar type, if you want the canonical type, you construct that.  The type checking is purely on canonical type, so it will pass either way.

Comment (mikeurbach) [reply include/circt/Dialect/HW/HWTypes.h:47]: The type checking is indeed going to work with the canonical type where necessary, but I think it is best to have this method use `getInnerType` rather than `getCanonicalType`. This helper is called by parsing code, before going into the builder, so if we fully canonicalize the type, anything in the parsing code that looks at an ArrayType's element type may lose information if the element type is also an alias itself. I've updated this to use `getInnerType` to peel off the outermost alias only and preserve any nested type aliases during parsing.

**Comment by mikeurbach** [comment]
@lattner I'm coming back to this after our discussion in last week's ODM. One concrete suggestion I remember you made was this notion of `TypeAliasOr<ArrayType>` should more transparent, so the infrastructure can just work with `ArrayType`. So I moved the ODS definition of `ArrayType` into `ArrayTypeImpl`, and now `ArrayType` is defined as `TypeAliasOr<ArrayTypeImpl>` in ODS. All the ops that work with `ArrayType` are unchanged, at least in their ODS definition.

However, I think this doesn't fully cover what you were getting at before. For example, the parsing code needs still needs to be updated to parse a `TypeAliasOr<ArrayType>` and unpack it in C++ rather than parsing a plain `ArrayType`. Was your intent for it to be fully transparent such that we could just parse an `ArrayType`, and a type alias would automagically know how to parse as an `ArrayType`? I'm just trying to hone in on what the desired end state should be.

I think the current approach in this PR is about as far was we can go down this route of having a `TypeAliasOr` dialect type. I'm starting to think that a more profitable route would be to have the base `HWType` have the functionality to transparently work as either an alias or the base type. If I can make this functionality work there, it should just work for `ArrayType` and anything else that extends `HWType`.

When you get a chance, please let me know if I'm at least understanding the goal. Again, this discussion is only with regard to concrete subclasses of `HWType`, the dialect types like `HWValueType`, `HWNonInOutType`, etc. already support transparently seeing through aliases with https://github.com/llvm/circt/pull/1076.

**Comment by mikeurbach** [comment]
> Inspired by the Clang docs, I think what I'm proposing in the paragraph above would instead be `type.cast<HWType>().getAsArrayType().getSize()`

I went ahead and moved to this implementation in the previous commit. The Clang internals documentation now makes sense to me, and I think this is a decent realization of those ideas in MLIR.

> The only hard part here is remembering not to use the isa/cast/dyn_cast operations.

> This structure is somewhat mystical, but after meditating on it, it will make sense to you :)

@lattner when you get a chance, I think this is ready for review.

**Comment by mikeurbach** [comment]
Ack... not sure why the test failed, it worked locally... will track it down.

File: include/circt/Dialect/HW/HWTypes.h
Comment (teqdruid) [inline include/circt/Dialect/HW/HWTypes.h:36]: Why not provide a templatized version of this and the get function below in addition to or instead of this? `bool isActually<T>()`?

**Review by teqdruid — APPROVED** [review]
This approach looks fine to me but I have no previous experience representing typedefs in an IR like @lattner.

File: include/circt/Dialect/HW/HWTypes.h
Comment (mikeurbach) [reply include/circt/Dialect/HW/HWTypes.h:36]: I was just following the method naming as described in the Clang internals doc, but it should be possible to template this by the desired type...

File: include/circt/Dialect/HW/HWAggregates.td
Comment (mikeurbach) [inline include/circt/Dialect/HW/HWAggregates.td:134]: `type_cast<ArrayType>($self).getElementType()`

File: include/circt/Dialect/HW/HWTypes.h
Comment (mikeurbach) [reply include/circt/Dialect/HW/HWTypes.h]: After moving away from the Clang-style, these helpers are indeed templated.

**Comment by mikeurbach** [comment]
@lattner when you get a chance, please let me know what you think here. I added the `type_cast` thing like we discussed in the ODM today.

A couple points I wasn't sure about:
* Spelling: the helpers are currently named `type_isa` and `type_cast`. This is in violation of our clang-tidy expectation, but in line with `dyn_cast` and `dyn_cast_or_null`. I was thinking since the LLVM `dyn_cast`, etc. are sort of "operators" LLVM is adding to C++, they have different rules? Regardless, it seemed appropriate to snake case here to be consistent, but I'm not too tied to it either way.
* Location: the helpers are added to the HW dialect right now, within the `circt::hw` namespace. To me, it seems like these are "operators" we're adding for working with the HW type system, and it makes sense to have them in `HWTypes.h` within the `circt::hw` namespace. Did you have something else in mind?

**Comment by mikeurbach** [comment]
@lattner I posted about potentially discussing this at an MLIR ODM, but in the short term, what do you think about moving forward with the current approach and then iterating?

To catch you up, we already have the `typescope` and `typedecl` ops, and a first draft of the `typealias` type. We previously gave up on resolving the inner type at parse time, and just require it to be passed in to the typealias builder:

https://github.com/llvm/circt/blob/29af376335f820dd80b86de850e5763487e4fb1b/include/circt/Dialect/HW/HWTypes.td#L163-L166

I'm happy to keep reworking this stuff as we get more clarity about the best way forward, but this PR (or something like it) should unlock the use cases @teqdruid and I are interested in (coming from the PyCDE frontend and emitting readable System Verilog for complex array and struct types).

**Comment by mikeurbach** [comment]
@lattner after last week's ODM, it seemed like we didn't come to any concrete conclusion about a way forward here. What do you think about the previous comment, i.e. doing something like this PR for now to get a basic typedef working, and then improving as the community continues to discuss? The internal project I've been working on with @teqdruid is going to need some kind of typedef support sooner rather than later.

I will find some time to poke around and see about this idea of resolving the symbol reference at parse time, but if I do get that to work eventually, I think we'd still want something along these lines anyway.

File: lib/Dialect/HW/HWOps.cpp
Comment (teqdruid) [inline lib/Dialect/HW/HWOps.cpp:834]: It is unfortunate that `parseType` doesn't for aliased types. Further motivation for getting typedefs supported upstream.

File: lib/Dialect/HW/HWOps.cpp
Comment (teqdruid) [inline lib/Dialect/HW/HWOps.cpp:898]: I'd suggest creating a `type_dyn_cast` (maybe just converting `type_cast` over) and using it here and above.

**Review by teqdruid — APPROVED** [review]
Looks about right to me!

File: lib/Dialect/HW/HWOps.cpp
Comment (mikeurbach) [reply lib/Dialect/HW/HWOps.cpp:898]: Fair, I didn't see these use cases at first, but a dyn_cast would be appropriate here. I'll keep all three--there are some places where ODS guarantees the type_cast will succeed and we shouldn't need to use a dyn_cast.

**Review by teqdruid — APPROVED** [review]
LGTM!

**Comment by mikeurbach** [comment]
Thanks. Just waiting for the tests to pass, and then I'm going to merge this. As I mentioned above, we'll need something like this to see through type aliases, however we end up parsing out the inner type. Since this has been open for more than three months, and gone through three significant iterations after ODM discussion, I think this is a decent starting point. Of course, I'll incorporate any post-commit feedback and discussion in the community, but I think it's time to have something that works in-tree.

**Pull request merged**
